File: programmers/타겟 넘버.py
"""
코딩테스트 연습 > 깊이/너비 우선 탐색(DFS/BFS)
"""


def dfs(numbers, i, sum, target):
    if i == len(numbers):
        if sum == target:
            return 1
        else:
            return 0
    else:
        return dfs(numbers, i+1, sum+numbers[i], target) + dfs(numbers, i+1, sum-numbers[i], target)


def solution(numbers, target):
    return dfs(numbers, 0, 0, target)


print(solution([1, 1, 1, 1, 1], 3))  # 5
print(solution([4, 1, 2, 1], 4))  # 2

File: programmers/타겟 넘버.kt
class Solution {
    fun solution(numbers: IntArray, target: Int): Int {
        var count = 0

        fun dfs(currentSum: Int, index: Int) {
            if (index == numbers.size) {
                if (currentSum == target) {
                    count += 1
                }
            } else {
                val num = numbers[index]
                dfs(currentSum + num, index + 1)
                dfs(currentSum - num, index + 1)
            }
        }

        dfs(0, 0)
        println(count)
        return count
    }
}
